### ISSUE TITLE
`composer up` stuck in infinite loop

### ISSUE THREAD

**Issue by jderusse** [composer/composer#9503]
My `composer.json`:

```json
{
    "require": {
        "symfony/framework-bundle": "^5.0",
        "phpstan/phpstan": "^0.11"
    }
}
```

Output of `composer diagnose`:

```
Checking composer.json: WARNING
No license specified, it is recommended to do so. For closed-source software you may use "proprietary" as license.                                                                                                  
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com oauth access: OK
Checking disk free space: OK
Checking pubkeys:
Tags Public Key Fingerprint: 57815BA2 7E54DC31 7ECC7CC5 573090D0  87719BA6 8F3BB723 4E5D42D0 84A14642
Dev Public Key Fingerprint: 4AC45767 E5EC2265 2F0C1167 CBBB8A2B  0C708369 153E328C AD90147D AFE50952
OK
Checking composer version: You are not running the latest snapshot version, run `composer self-update` to update (2.0.7 => 9e279740cd4dea381a77cc48af617a91d835408e)                                                
Composer version: 2.0.7
PHP version: 7.3.19
PHP binary path: /usr/bin/php7.3
OpenSSL version: OpenSSL 1.1.1d  10 Sep 2019
cURL version: 7.64.0 libz 1.2.11 ssl OpenSSL/1.1.1d
zip extension: OK
```

When I run this command:

```
composer up -vvv
```

I get the following output:

```
....
Something's changed, looking at all rules again (pass #157)
```

And I expected this to happen:

I expect composer either found a matching version of failed with an error.
 
nb: Sorry @Seldaek  I don'thave time to dig into this at the moment... :( I'll have a look this week-end

**Comment by stof** [comment]
phpstan 0.11 has requirements on `symfony/console` `~3.2 || ~4.0` (and same for `symfony/finder`), so this is a case of unsolvable dependencies. And such cases are the worst case of a SAT solver, which might have to try all potential combinations of packages before deciding that none of the work (#9261 will also help there, as it will reduce a lot the number of combinations in the solver).

Note that the fix is easy: use an uptodate and maintained phpstan (i.e. 0.12)

**Comment by Seldaek** [comment]
157 passes  is not that many, I've seen this go up to thousands before a problem output came out :P But usually when it reaches that point, nothing good is going to come out of it, and as @stof mentioned it's unsolvable here..

**Comment by stof** [comment]
note that the loop is still finite (because the combinations of versions of packages involved in the solver are a finite set). But it is very large (probably something like `number of packages ** number of versions per packages` and Symfony packages have 501 versions currently)

**Comment by jderusse** [comment]
OK... indeed, I didn't waited enough.. It has been solved after 1 minute and `pass #2014`... 
I didn't expect that for a such "simple" composer.json

> phpstan 0.11 has requirements on symfony/console ~3.2 || ~4.0 (and same for symfony/finder), so this is a case of unsolvable dependencies. And such cases are the worst case of a SAT solver

This were a minimal reproducer, we have the exact same trouble with a matching constraint

```json
{
    "require": {
        "symfony/framework-bundle": "^4.0|^5.0",
        "phpstan/phpstan": "^0.11"
    }
}
````

> Note that the fix is easy: use an uptodate and maintained phpstan

The issue is about performance issue with a specific set of constraints. I thought it was infinite, but, has you pointed it, I was wrong, it's just super slow. 
With a "real world" project and ~20 root requirements, it took me more than 3 minutes and `3813 passes`

Of course, When we changes deps, the symptom disappear, but IMHO the performance issue is still here.

**Comment by stof** [comment]
@jderusse and as I said, #9261 is precisely the in-progress work to solve that performance issue, by limiting the number of versions entering the solver for each package (see my comment above for the impact it has on the worst case complexity of the SAT solver)
